### ISSUE TITLE
Cannot read property 'type' of null error in no-array-index-key.js

### ISSUE THREAD

**Issue by alaycock** [yannickcr/eslint-plugin-react#1242]
Version 7.0.1
The full stack trace can be found below.

![screen shot 2017-06-06 at 1 20 23 pm](https://user-images.githubusercontent.com/894797/26847609-ebf2295a-4aba-11e7-8f88-9989b3044691.png)

The issue occurs when I begin to add the key to this code:

          this.state.array.map((arrayItem, index) => (
            <div>
              <button onClick={() => this.delete(index)}>Delete this thing</button>
            </div>
          ))

If I begin to enter the `key` attribute:

          this.state.array.map((arrayItem, index) => (
            <div key>
              <button onClick={() => this.delete(index)}>Delete this thing</button>
            </div>
          ))

An error in the console is triggered, although I wasn't finished writing the new line.

The issue appears to be [on this line](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/no-array-index-key.js#L180), and I suspect it could be corrected altering the line to be `if (!value || value.type !== 'JSXExpressionContainer') {`, but I'm not sure about the implication of making that change.

**Comment by ljharb** [comment]
So, setting aside "errors while you're typing", which isn't something that needs supporting, it looks like `<div key />`inside a `.map` throws this error?

**Comment by alaycock** [comment]
Exactly. It looks like the variable which holds the value is null, which throws the TypeError.

**Comment by jseminck** [comment]
Confirmed by adding a test case:

```js
  1) no-array-index-key invalid foo.map((bar, i) => <Foo key />):
     TypeError: Cannot read property 'type' of null
```

But... should it be valid, or invalid? According to this specific rule I guess it should be valid, since it's not assigning the index to the key... but then again, just having `foo.map((bar, i) => <Foo key />)` seems wrong?

**Comment by ljharb** [comment]
In that case, it's setting the `key` to `true` - which is not a use of the index. Separately, it's making the key be the same, which defeats the purpose - but ensuring the key is unique across items is something that React could check at runtime, not something a linter rule could reliably enforce.
